docs/projects/merge-sort/_posts/2019-01-24-python.md

```diff
@@ -0,0 +1,340 @@
+calls itself within the function, in this case the `merge()` function calling the `merge()`
+function again.
+
+Since `xs[0]` is less than `y[0]` we know `xs[0]` must be
+the smallest item in either list, as both lists are already sorted. So we then take `xs[0]` out
+of the first list we convert it into a list `[xs[0]]` because we want to concatenate it with a
+sorted list. The `+` operator can be used to concatenate two lists together.
+
+We call `merge(xs[1:], ys)` to sort the remaining items in both lists. Where `xs[1:]` is the list
+`xs` without the first element. This is called index splicing in Python, you can learn more about it
+[here](https://www.pythoncentral.io/how-to-slice-listsarrays-and-tuples-in-python/).
+We don't include the first element of `xs` because we already know it's the smallest element in
+both lists. The final line `return [ys[0]] + merge(xs, ys[1:])` is the same as line above except
+in this case `ys[0]` is smaller than `xs[0]` hence we use `ys[0]` at the start of the list
+concatenation instead of `xs[0]`. The reason this isn't surrounded by an `else` statement is
+because of the return statement nothing will be executed after it.
+
+Let's take a look at an example, where `xs = [3, 4]` and `ys = [1, 10]`
+
+Merge 1:
+
+* xs = [3, 4]
+* ys = [1, 10]
+* len(xs) = 2 and 2 > 0
+* len(ys) = 2 and 2 > 0
+* `xs[0] = 3` and `ys[0] = 1` and 3 > 1
+* [1] + merge([3, 4], [10])
+
+Merge 2:
+
+* xs = [3, 4]
+* ys = [10]
+* len(xs) = 2 and 2 > 0
+* len(ys) = 1 and 1 > 0
+* `xs[0] = 3` and `ys[0] = 4` and 3 < 4
+* [3] + merge([4], [10])
+
+Merge 3
+
+* xs = [4]
+* ys = [10]
+* len(xs) = 1 and 1 > 0
+* len(ys) = 1 and 1 > 0
+* `xs[0] = 4` and `ys[0] = 10` and 4 < 10
+* [4] + merge([], [10])
+
+Merge 4
+
+* xs = []
+* ys = [10]
+* len(xs) = 0 and 0 <= 0
+* return `[10]`
+
+Going backwards
+
+* [4] + [10] = [4, 10] (Merge 3)
+* [3] + [4, 10] = [3, 4, 10] (Merge 2)
+* [1] + [3, 4, 10] = [1, 3, 4, 10] (Merge 1)
+
+### Merge Sort
+
+```python
+def merge_sort(xs):
+    def sort(xs):
+        if len(xs) <= 0:
+            return []
+        if len(xs) == 1:
+            return xs
+        return sort([merge(xs[0], xs[1])] + sort(xs[2:]))
+    split_xs = [[x] for x in xs]
+    return sort(split_xs)[0]
+```
+
+This is the function that actually sorts our entire list. 
+
+```python
+    def merge_sort(xs):
+        def sort(xs):
+```
+
+This is known as closure nested functions (a function within another function).
+The nested function can access variables of the enclosing scope. They can read
+variable but cannot edit them, in the example above it's the `xs` variable our
+unsorted list.
+
+The `sort()` function only exists within the scope of `merge_sort()` but other than that
+is works exactly the same as a normal function. 
+
+```python
+split_xs = [[x] for x in xs]
+return sort(split_xs)[0]
+````
+
+So in this example we execute the above lines first you can ignore the `sort()` function until
```

Comment:
Again, I'm not sure where to break these two sentence. Take a look and let me know!

```suggestion
In this example, we execute the above lines. First. you can ignore the `sort()` function until
```